sbndqm/dqmAnalysis/PMT/ICARUS/BeamTimingStreams_module.cc

```diff
@@ -0,0 +1,335 @@
+  , m_ADC_threshold{ pset.get<short int>("ADCthreshold", 200)}
+  , m_OpticalTick{ 0.002 }
+{
+  // Configure the redis metrics 
+  sbndaq::GenerateMetricConfig( m_metric_config );
+}
+
+//------------------------------------------------------------------------------------------------------------------
+
+template<typename T> T sbndaq::BeamTimingStreams::Median( std::vector<T> data ) const {
+
+  std::nth_element( data.begin(), data.begin() + data.size()/2, data.end() );
+  return data[ data.size()/2 ];
+
+}  
+
+//------------------------------------------------------------------------------------------------------------------
+
+template<typename T> size_t sbndaq::BeamTimingStreams::getMinBin(std::vector<T> const& vv, size_t startElement, size_t endElement ){
+
+  auto minel = std::min_element( vv.begin()+startElement, vv.begin()+endElement );
+  size_t minsample = std::distance( vv.begin()+startElement, minel );
+  return minsample;
+
+}
+
+//------------------------------------------------------------------------------------------------------------------
+
+template<typename T> size_t sbndaq::BeamTimingStreams::getMaxBin(std::vector<T> const& vv, size_t startElement, size_t endElement){
+
+  auto maxel = std::max_element( vv.begin()+startElement, vv.begin()+endElement );
+  size_t maxsample = std::distance( vv.begin()+startElement, maxel );
+  return maxsample;
+
+} 
+
+//------------------------------------------------------------------------------------------------------------------
+
+template<typename T> size_t sbndaq::BeamTimingStreams::getStartSample( std::vector<T> const& vv, T thres ){
+    
+  // get the pulse minimum (undershoot of the signal)
+  size_t minbin = getMinBin( vv, 0, vv.size() );
+
+  //search only a cropped region of the waveform backward from the min
+  size_t maxbin =  minbin-20; 
+
+  // Now we crawl betweem maxbin and minbin and we stop when:
+  // bin value > ( maxbin value - minbin value )*0.2
+  size_t startbin = maxbin;
+  auto delta = vv[maxbin]-vv[minbin];
+ 
+  if( delta < thres ) // just noise
+    return 0; //return first bin 
+
+  for( size_t bin=maxbin; bin<minbin; bin++ ){
+    auto val = vv[maxbin]-vv[bin];
+    if( val >= 0.2*delta ){  // 20%
+      startbin = bin - 1;
+      break;
+    }
+  }
+
+  if( startbin < maxbin ){
+    startbin=maxbin;
+  }
+
+  return startbin;
+}
+
+//------------------------------------------------------------------------------------------------------------------
+
+std::string sbndaq::BeamTimingStreams::getGateName(int source){
+  
+  switch(source){
+    case daq::TriggerGateTypes::BNB:
+      return "BNB";
+    case daq::TriggerGateTypes::NuMI:
+      return "NUMI";
+    default:
+      return "Offbeam";
+  }
+  
+}
+
+//------------------------------------------------------------------------------------------------------------------
+
+bool sbndaq::BeamTimingStreams::isSpecialChannel(int pmtID, std::map<size_t,std::string> channel_map, int &crateID){
```

Comment:
Pass the map by constant reference:
```suggestion
bool sbndaq::BeamTimingStreams::isSpecialChannel(int pmtID, std::map<size_t,std::string> const& channel_map, int &crateID){
```